samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerUtil.java

```diff
@@ -411,4 +438,292 @@ public static Set<String> getSideInputStoreNames(
+    // ContainerStorageManager we always start the changelog consumer here in case it is required
+    storeConsumers.values().stream().distinct().forEach(SystemConsumer::start);
+
+    return restoreAllTaskInstances(taskRestoreManagers, taskCheckpoints, taskBackendFactoryToStoreNames, jobContext,
+        containerModel, samzaContainerMetrics, checkpointManager, config, taskInstanceMetrics, executor, loggerStoreDir,
+        forceRestoreTasks);
+  }
+
+  /**
+   * Restores all TaskInstances and returns a future for each TaskInstance restore. If this restore fails with DeletedException
+   * it will retry the restore with getDeleted flag, get all the blobs, recreate a new checkpoint in blob store and
+   * write that checkpoint to checkpoint topic.
+   * @param taskRestoreManagers map of TaskName to factory name to TaskRestoreManager map.
+   */
+  public static List<Future<Void>> restoreAllTaskInstances(
+      Map<TaskName, Map<String, TaskRestoreManager>> taskRestoreManagers, Map<TaskName, Checkpoint> taskCheckpoints,
+      Map<TaskName, Map<String, Set<String>>> taskBackendFactoryToStoreNames, JobContext jobContext,
+      ContainerModel containerModel, SamzaContainerMetrics samzaContainerMetrics, CheckpointManager checkpointManager,
+      Config config, Map<TaskName, TaskInstanceMetrics> taskInstanceMetrics, ExecutorService executor,
+      File loggedStoreDir, Set<String> forceRestoreTask) {
+
+    List<Future<Void>> taskRestoreFutures = new ArrayList<>();
+
+    // Submit restore callable for each taskInstance
+    taskRestoreManagers.forEach((taskInstanceName, restoreManagersMap) -> {
+      // Submit for each restore factory
+      restoreManagersMap.forEach((factoryName, taskRestoreManager) -> {
+        long startTime = System.currentTimeMillis();
+        String taskName = taskInstanceName.getTaskName();
+        LOG.info("Starting restore for state for task: {}", taskName);
+
+        CompletableFuture<Void> restoreFuture;
+        if (forceRestoreTask.contains(taskName) && taskRestoreManager instanceof BlobStoreRestoreManager) {
+          restoreFuture = ((BlobStoreRestoreManager) taskRestoreManager).restore(true);
+        } else {
+          restoreFuture = taskRestoreManager.restore();
+        }
+
+        CompletableFuture<Void> taskRestoreFuture = restoreFuture.handle((res, ex) -> {
+          updateRestoreTime(startTime, samzaContainerMetrics, taskInstanceName);
+
+          if (ex != null) {
+            if (isUnwrappedExceptionDeletedException(ex)) {
+              LOG.warn(
+                  "Restore state for task: {} received DeletedException. Reattempting with getDeletedBlobs enabled",
+                  taskInstanceName.getTaskName());
+
+              // Try to restore with getDeleted flag
+              CompletableFuture<Void> future =
+                  restoreTaskStoresAndCreateCheckpointWithGetDeleted(taskInstanceName, taskCheckpoints, checkpointManager,
```

Comment:
Can these be two separate methods?
restoreWithGetDeleted (returns checkpoint) and update/writeCheckpoint?

Reply:
Renamed it to better reflect the intent. Please let me know if that's better or if I should split.

Reply:
LGTM, thanks.